compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/StringTrimLowering.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.common.lower
+
+import org.jetbrains.kotlin.backend.common.CommonBackendContext
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.lower.matchers.SimpleCalleeMatcher
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.isString
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+import org.jetbrains.kotlin.name.FqName
+
+val computeStringTrimPhase = makeIrFilePhase(
+    ::StringTrimLowering,
+    name = "StringTrimLowering",
+    description = "Compute trimIndent and trimMargin operations on constant strings"
+)
+
+class StringTrimLowering(val context: CommonBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid(this)
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        return when {
+            trimIndentMatcher(expression) -> maybeComputeTrimIndent(expression)
+            trimMarginMatcher(expression) -> maybeComputeTrimMargin(expression)
+            else -> super.visitCall(expression)
+        }
+    }
+
+    private fun maybeComputeTrimIndent(call: IrCall): IrExpression {
+        val receiverString = call.extensionReceiver!!.toConstantString() ?: return call
+        val newString = receiverString.trimIndent()
+        return IrConstImpl.string(call.startOffset, call.endOffset, call.type, newString)
+    }
+
+    private fun maybeComputeTrimMargin(call: IrCall): IrExpression {
+        val receiverString = call.extensionReceiver!!.toConstantString() ?: return call
+
+        val prefixArgument = call.getValueArgument(0)
+        val newString = if (prefixArgument != null) {
+            val prefixString = prefixArgument.toConstantString() ?: return call
+            receiverString.trimMargin(prefixString)
+        } else {
+            receiverString.trimMargin()
+        }
+
+        return IrConstImpl.string(call.startOffset, call.endOffset, call.type, newString)
+    }
+
+    companion object {
+        private fun IrExpression.toConstantString(): String? {
+            return when (this) {
+                is IrConst<*> -> IrConstKind.String.valueOf(this)
+                is IrStringConcatenation -> toConstantString()
```

Comment:
I feel like this should be its own lowering phase for all platforms. Currently this phrase gets IR like:
```
$receiver: STRING_CONCATENATION type=kotlin.String
  CONST String type=kotlin.String value="\n        |Hello,\n        |"
  CONST String type=kotlin.String value="Hello,\n        |World"
  CONST String type=kotlin.String value="\n        |World\n    "
```
when it really should only have to care about / check for `IrConst<*>`.

Reply:
Just found https://github.com/JetBrains/kotlin/pull/2274 which seems like it should partially handle this, although it omits support for all constants types.